File: src/opencode_cli_mcp/tools/runs.py
from typing import Annotated

from pydantic import Field

from opencode_cli_mcp.job_store import cancel_job, get_job, list_jobs


async def opencode_get_run_status(
    job_id: Annotated[str, Field(description="Job ID from opencode_run_agent")],
) -> dict:
    """Poll the status of a background agent run. Returns stdout/stderr accumulated so far, status, and exit code."""  # noqa: E501

    job = await get_job(job_id)
    if not job:
        return {"success": False, "message": f"Job '{job_id}' not found", "data": {}}

    return {
        "success": True,
        "message": f"Job {job['status']}",
        "data": {
            "job_id": job_id,
            "status": job["status"],
            "stdout": job["stdout"],
            "stderr": job["stderr"],
            "exit_code": job["exit_code"],
            "error": job.get("error"),
            "created_at": job["created_at"],
            "completed_at": job["completed_at"],
        },
    }


async def opencode_list_runs(
    limit: Annotated[int, Field(description="Max jobs to return", ge=1, le=100)] = 20,
) -> dict:
    """List all recent agent runs with their status and exit codes."""  # noqa: E501

    jobs = await list_jobs(limit=limit)
    return {
        "success": True,
        "message": f"Found {len(jobs)} runs",
        "data": {
            "runs": [
                {
                    "job_id": j["job_id"],
                    "prompt": j["prompt"],
                    "status": j["status"],
                    "exit_code": j["exit_code"],
                    "created_at": j["created_at"],
                    "completed_at": j["completed_at"],
                }
                for j in jobs
            ]
        },
    }


async def opencode_cancel_run(
    job_id: Annotated[str, Field(description="Job ID to cancel")],
) -> dict:
    """Cancel a running or queued agent run."""  # noqa: E501

    ok = await cancel_job(job_id)
    if not ok:
        return {"success": False, "message": f"Cannot cancel job '{job_id}' (not running or not found)", "data": {}}  # noqa: E501
    return {"success": True, "message": f"Cancelled job '{job_id}'", "data": {"job_id": job_id}}

File: .opencode/tools/runs.ts
import { tool } from "@opencode-ai/plugin"

const API = "http://127.0.0.1:10951/api"

export const list = tool({
  description: "List recent agent runs dispatched by the MCP server",
  args: {
    limit: tool.schema.number().describe("Max runs to return (default 10)").default(10),
  },
  async execute(args) {
    const resp = await fetch(`${API}/runs`)
    const { data } = await resp.json() as { success: boolean; data: { runs: Array<{ job_id: string; prompt: string; status: string; exit_code: number | null; created_at: number }> } }
    if (!data?.runs?.length) return "No agent runs found."
    const limited = data.runs.slice(0, args.limit)
    return limited.map(r => {
      const ts = new Date(r.created_at * 1000).toISOString().slice(0, 19)
      return `[${r.status}] ${r.job_id} - "${r.prompt.slice(0, 60)}" ${r.exit_code !== null ? `(exit ${r.exit_code})` : ""} ${ts}`
    }).join("\n")
  },
})

export const status = tool({
  description: "Get status of a specific agent run by job ID",
  args: {
    jobId: tool.schema.string().describe("The job ID to check"),
  },
  async execute(args) {
    const resp = await fetch(`${API}/runs/${args.jobId}`)
    if (!resp.ok) return `Run '${args.jobId}' not found.`
    const { data } = await resp.json() as { success: boolean; data: { run: { job_id: string; prompt: string; status: string; stdout: string; stderr: string; exit_code: number | null; created_at: number; completed_at: number | null } } }
    const r = data.run
    const ts = new Date(r.created_at * 1000).toISOString().slice(0, 19)
    let out = `Run ${r.job_id}: ${r.status}\nPrompt: "${r.prompt}"\nStarted: ${ts}`
    if (r.completed_at) out += `\nCompleted: ${new Date(r.completed_at * 1000).toISOString().slice(0, 19)}`
    if (r.exit_code !== null) out += `\nExit code: ${r.exit_code}`
    if (r.stdout) out += `\n\nOutput:\n${r.stdout.slice(0, 2000)}`
    if (r.stderr) out += `\n\nStderr:\n${r.stderr.slice(0, 1000)}`
    return out
  },
})
